core/src/main/java/org/apache/lucene/search/grouping/CollapsingDocValuesSource.java

```diff
@@ -0,0 +1,215 @@
+            super(field);
+        }
+
+        @Override
+        public BytesRef get(int doc) {
+            if (docsWithField.get(doc)) {
+                return values.get(doc);
+            } else {
+                return null;
+            }
+        }
+
+        @Override
+        public BytesRef copy(BytesRef value, BytesRef reuse) {
+            if (value == null) {
+                return null;
+            }
+            if (reuse != null) {
+                reuse.bytes = ArrayUtil.grow(reuse.bytes, value.length);
+                reuse.offset = 0;
+                reuse.length = value.length;
+                System.arraycopy(value.bytes, value.offset, reuse.bytes, 0, value.length);
+                return reuse;
+            } else {
+                return BytesRef.deepCopyOf(value);
+            }
+        }
+
+        @Override
+        public void setNextReader(LeafReader reader) throws IOException {
+            DocValuesType type = getDocValuesType(reader, field);
+            if (type == null) {
+                values = DocValues.emptySorted();
+                docsWithField = new Bits.MatchNoBits(reader.maxDoc());
+                return ;
+            }
+            docsWithField = DocValues.getDocsWithField(reader, field);
+            switch (type) {
+                case SORTED:
+                    values = DocValues.getSorted(reader, field);
+                    break;
+
+                case SORTED_SET:
+                    final SortedSetDocValues sorted = DocValues.getSortedSet(reader, field);
+                    values = DocValues.unwrapSingleton(sorted);
+                    if (values == null) {
+                        values = new SortedDocValues() {
+                            @Override
+                            public int getOrd(int docID) {
+                                sorted.setDocument(docID);
+                                int ord = (int) sorted.nextOrd();
+                                if (sorted.nextOrd() != SortedSetDocValues.NO_MORE_ORDS) {
+                                    throw new IllegalStateException("failed to collapse " + docID +
+                                        ", the collapse field must be single valued");
+                                }
+                                return ord;
+                            }
+
+                            @Override
+                            public BytesRef lookupOrd(int ord) {
+                                return sorted.lookupOrd(ord);
+                            }
+
+                            @Override
+                            public int getValueCount() {
+                                return (int) sorted.getValueCount();
+                            }
+                        };
+                    }
+                    break;
+
+                default:
+                    throw new IllegalStateException("unexpected doc values type "
+                        + type + "` for field `" + field + "`");
```

Comment:
do we need to handle NONE? I'm wondering this could happen if there is a segment that does not have any values for that field